Clarify space partitioning due to multi-node (#564)
timescale/docs.timescale.com-content: api.md

converted to a hypertable (via `create_hypertable`), but must similarly
be run only on an empty hypertable.

> **Space partitions**: Using space partitions is important for efficient scale out 
> in [multi-node setup][multinode]. In the case of 
> single-node the use of additional partitioning is a very
> specialized use case and most users will not need it.

Reviewer: ```suggestion
**Space partitions**: Using space partitions is highly recommended
for [distributed hypertables](#create_distributed_hypertable) to achieve
efficient scale-out performance. For [regular hypertables](#create_hypertable)
that exist only a single node, the use of additional partitioning is a
specialized use case and not recommended for most users.
```

Space partitions use hashing: Every distinct item is hashed to one of
*N* buckets.  Remember that we are already using (flexible) time
intervals to manage chunk sizes; the main purpose of space

> partitioning is to enable parallel utilization of multiple data nodes or 
> parallel I/O on a single node to the same time interval.

Reviewer: ```suggestion
partitioning is to enable parallelization across multiple 
data nodes (in the case of distributed hypertables) or
across multiple disks within the same time interval
(in the case of single-node deployments).
```

> ### Space partitions for multi-node

Reviewer: ```suggestion
### Parallelizing queries across multiple data nodes [](add_dimension-multi_node)
```

> Space partitioning allows to parallelize inserts into different data nodes
> and increase the ingesting rate. Select queries will also benefit when 
> operations, which reduce data, e.g., conditions in WHERE clause or
> aggregate operators, are pushed to data nodes. However, some operations 
> are not pushed, e.g., joins or non-parallelizable aggregates, and thus
> all needed data will be transferred to the access node, which might be expensive.
> 
> It is recommended to set the number of partitions to the number of attached data nodes.

Reviewer: ```suggestion
In a distributed hypertable, space partitioning enables inserts to be
parallelized across data nodes, even while the inserted rows share
timestamps from the same time interval, and thus increases the ingest rate.
Query performance also benefits by being able to parallelize queries 
across nodes, particularly when full or partial aggregations can be
"pushed down" to data nodes (e.g., as in the query
`avg(temperature) FROM conditions GROUP BY hour, location`
when using `location` as a space partition). Please see our
[best practices about partitioning in distributed hypertables][distributed-hypertable-partitioning-best-practices]
for more information.
```

> ### Parallel I/O on a single node

Reviewer: ```suggestion
### Parallelizing disk I/O on a single node [](add_dimension-single_node)
```

converted to a hypertable (via `create_hypertable`), but must similarly
be run only on an empty hypertable.

**Space partitions**: Using space partitions is highly recommended
for [distributed hypertables](#create_distributed_hypertable) to achieve
efficient scale-out performance. For [regular hypertables](#create_hypertable)

> that exist only a single node, the use of additional partitioning is a

Reviewer 2: only on a?

Author: I improved this place.

specialized use case and not recommended for most users.

Space partitions use hashing: Every distinct item is hashed to one of
*N* buckets.  Remember that we are already using (flexible) time
intervals to manage chunk sizes; the main purpose of space
partitioning is to enable parallelization across multiple 
data nodes (in the case of distributed hypertables) or
across multiple disks within the same time interval
(in the case of single-node deployments).

> ### Parallelizing queries across multiple data nodes [](add_dimension-multi_node)

Reviewer 2: Curious if we should move this sections somewhere else, like performance or optimization

Author: I suggest to discuss and consider this outside this PR.

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

> First, create a multinode with two data nodes, then convert table `conditions` to 
> a distributed hypertable with just time partitioning on column `time`, 
> finally add a space partitioning on `location` with two partitions:

Reviewer: ```suggestion
```suggestion
Now in a multi-node example for distributed hypertables, 
setup your cluster to have one access node and two data nodes, 
and configure your access node for access to the data nodes.
Then, convert your table `conditions` to 
a distributed hypertable with just time partitioning on column `time`, 
and finally add a space partitioning dimension on `location`.  
By default, the `number_partitions` will be set to the number of data nodes
in the distributed hypertable `conditions`.
```

```sql
SELECT add_data_node('data_node1', host => 'node1_host_address');
SELECT add_data_node('data_node2', host => 'node2_host_address');
SELECT create_distributed_hypertable('conditions', 'time');

> SELECT add_dimension('conditions', 'location', number_partitions => 2);

Reviewer: ```suggestion
SELECT add_dimension('conditions', 'location');
```

Author: @mfreed It doesn't work in this way. `add_dimension` in you suggestion will return an error that `number_partitions` (or time interval) must be provided.

Author: The default exists only for `create_distributed_hypertable`.

Reviewer: Sounds good.  Curious -> is this fundamental or just an API thing we didn't get around to updating?

Author: Being discussed outside the PR.

SELECT add_dimension('conditions', 'device_id', number_partitions => 2, if_not_exists => true);
```

Now in a multi-node example for distributed hypertables with a cluster 
of one access node and two data nodes, configure the access node for 
access to the two data nodes. Then, convert table `conditions` to 
a distributed hypertable with just time partitioning on column `time`, 
and finally add a space partitioning dimension on `location`
with two partitions (as the number of the attached data nodes).

```sql
SELECT add_data_node('dn1', host => 'dn1.example.com');

> SELECT add_data_node('dn2', host => 'dn1.example.com');

Reviewer 2: `dn2` is using `dn1.example.com` host

Author: Thank you! Fixed.

[postgres-lock]: https://www.postgresql.org/docs/current/sql-lock.html
[migrate-from-postgresql]: /getting-started/migrating-data
[memory-units]: https://www.postgresql.org/docs/current/static/config-setting.html#CONFIG-SETTING-NAMES-VALUES

> [multinode]: /getting-started/setup-multi-node-basic

Reviewer: ```suggestion
[multinode]: /getting-started/setup-multi-node-basic
[distributed-hypertable-partitioning-best-practices]: /using-timescaledb/distributed-hypertables#partitioning-best-practices
```
